wikidatarefisland/pipes/item_statistical_matcher_pipe.py

```diff
@@ -0,0 +1,93 @@
+
+    def flow(self, potential_match):
+        """
+        Observe and take notes of items passing by
+
+        This pipe doesn't produce any match on its own
+        but it only builds mapping of most used items and their values in the scraped data.
+        For example if a website uses "(Germany)" as the value of nationality of a person
+        and we have "Q183" as that value and it gets repeated more than N times,
+        we consider these two values to be the same.
+
+        :param potential_match: a line of output of pipe2
+        """
+        if potential_match['statement']['datatype'] != 'wikibase-item':
+            return [potential_match]
+        item_id = potential_match['statement']['value']['numeric-id']
+        ext_id_property = None
+        for pid in potential_match['reference']['referenceMetadata']:
+            if pid not in self.whitelisted_external_identifiers:
+                continue
+            ext_id_property = pid
+            break
+        if not ext_id_property:
+            return [potential_match]
+        extracted_data = potential_match['reference']['extractedData']
+        for value in extracted_data:
+            value = str(value)
+            if not value:
+                continue
+            pid = potential_match['statement']['pid']
+            if pid not in self.statistics[ext_id_property]:
+                self.statistics[ext_id_property][pid] = defaultdict(list)
+            self.statistics[ext_id_property][pid][str(value)].append(item_id)
+        return [potential_match]
+
+    def get_mapping(self):
+        """
+
+        :return: A dictionary of mapping of ext_id_property -> pid -> "value given by the ext id"
+            to numeric id of the item equivalent to it.
+            For example if external identifier "P1" provides "(Germany)" for
+            nationality property ("P27") and the value in Wikidata is Q183. The mapping will be:
+            mapping['P1']['P27']['(Germany)'] = 183
+        """
+        final_result = defaultdict(dict)
+        for ext_id_property in self.statistics:
+            for pid in self.statistics[ext_id_property]:
+                final_result[ext_id_property][pid] = self._analyze_property(ext_id_property, pid)
+
+        return final_result
+
+    def _analyze_property(self, ext_id_property, pid):
+        per_pid_result = {}
+        for value in self.statistics[ext_id_property][pid]:
+            items = self.statistics[ext_id_property][pid][value]
+            if len(items) > self.minimum_repetitions:
+                preferred_value = self.get_preferd_item_with_noise(items)
+                if preferred_value:
+                    per_pid_result[value] = self.get_preferd_item_with_noise(items)
+
+        return per_pid_result
+
+    def get_preferd_item_with_noise(self, items):
+        """
+        :type items: list
+        """
+        # Everything is the same, bail out.
+        if len(list(set(items))) == 1:
```

Comment:
I think we can use `len()` directly on `set()`s: https://docs.python.org/3.7/library/stdtypes.html#set

Reply:
oooh, thanks.